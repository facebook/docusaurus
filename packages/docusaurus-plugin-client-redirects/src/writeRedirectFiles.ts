/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import logger from '@docusaurus/logger';
import {normalizeUrl} from '@docusaurus/utils';

import createRedirectPageContent from './createRedirectPageContent';

import type {PluginContext, RedirectItem} from './types';

export type WriteFilesPluginContext = Pick<PluginContext, 'baseUrl' | 'outDir'>;

export type RedirectFile = {
  fileAbsolutePath: string;
  fileContent: string;
};

export function createToUrl(baseUrl: string, to: string): string {
  if (to.startsWith('/')) {
    return normalizeUrl([baseUrl, to]);
  }
  return to;
}

// Create redirect file path
// Make sure this path has lower precedence over the original file path when
// served by host providers!
// Otherwise it can produce infinite redirect loops!
//
// See https://github.com/facebook/docusaurus/issues/5055
// See https://github.com/facebook/docusaurus/pull/5085
// See https://github.com/facebook/docusaurus/pull/5102
function getRedirectFilePath(
  fromPath: string,
  trailingSlash: boolean | undefined, // Now unused, on purpose
): string {
  const fileName = path.basename(fromPath);
  const filePath = path.dirname(fromPath);
  // Edge case for https://github.com/facebook/docusaurus/pull/5102
  // If the redirect source path is /xyz, with file /xyz.html
  // We can't write the redirect file at /xyz.html/index.html because for Unix
  // FS, a file/folder can't have the same name "xyz.html"
  // The only possible solution for a redirect file is thus /xyz.html.html (I
  // know, looks suspicious)
  if (trailingSlash === false && fileName.endsWith('.html')) {
    return path.join(filePath, `${fileName}.html`);
  }
  // If the target path is /xyz, with file /xyz/index.html, we don't want the
  // redirect file to be /xyz.html, otherwise it would be picked in priority and
  // the redirect file would redirect to itself. We prefer the redirect file to
  // be /xyz.html/index.html, served with lower priority for most static hosting
  // tools
  return path.join(filePath, `${fileName}/index.html`);
}

export function toRedirectFiles(
  redirects: RedirectItem[],
  pluginContext: WriteFilesPluginContext,
  trailingSlash: boolean | undefined,
): RedirectFile[] {
  // Perf: avoid rendering the template twice with the exact same "props"
  // We might create multiple redirect pages for the same destination url
  // note: the first fn arg is the cache key!
  const createPageContentMemoized = _.memoize((toUrl: string) =>
    createRedirectPageContent({toUrl}),
  );

  const createFileMetadata = (redirect: RedirectItem) => {
    const fileRelativePath = getRedirectFilePath(redirect.from, trailingSlash);
    const fileAbsolutePath = path.join(pluginContext.outDir, fileRelativePath);
    const toUrl = createToUrl(pluginContext.baseUrl, redirect.to);
    const fileContent = createPageContentMemoized(toUrl);
    return {
      ...redirect,
      fileAbsolutePath,
      fileContent,
    };
  };

  return redirects.map(createFileMetadata);
}

export async function writeRedirectFile(file: RedirectFile): Promise<void> {
  try {
    // User-friendly security to prevent file overrides
    if (await fs.pathExists(file.fileAbsolutePath)) {
      throw new Error(
        'The redirect plugin is not supposed to override existing files.',
      );
    }
    await fs.outputFile(
      file.fileAbsolutePath,
      file.fileContent,
      // Hard security to prevent file overrides
      // See https://stackoverflow.com/a/34187712/82609
      {flag: 'wx'},
    );
  } catch (err) {
    logger.error`Redirect file creation error for path=${file.fileAbsolutePath}.`;
    throw err;
  }
}

export default async function writeRedirectFiles(
  redirectFiles: RedirectFile[],
): Promise<void> {
  await Promise.all(redirectFiles.map(writeRedirectFile));
}
