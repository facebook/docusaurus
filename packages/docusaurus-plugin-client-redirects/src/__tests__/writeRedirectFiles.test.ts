/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';

import writeRedirectFiles, {
  toRedirectFiles,
  createToUrl,
} from '../writeRedirectFiles';

// Test to fix toUrl bugs we had:
// - https://github.com/facebook/docusaurus/issues/3886
// - https://github.com/facebook/docusaurus/issues/3925
describe('createToUrl', () => {
  it('creates appropriate redirect urls', () => {
    expect(createToUrl('/', '/docs/something/else')).toBe(
      '/docs/something/else',
    );
    expect(createToUrl('/', '/docs/something/else/')).toBe(
      '/docs/something/else/',
    );
    expect(createToUrl('/', 'docs/something/else')).toBe('docs/something/else');
    expect(createToUrl('/', './docs/something/else')).toBe(
      './docs/something/else',
    );
    expect(createToUrl('/', 'https://docs/something/else')).toBe(
      'https://docs/something/else',
    );
  });

  it('creates appropriate redirect urls with baseUrl', () => {
    expect(createToUrl('/baseUrl/', '/docs/something/else')).toBe(
      '/baseUrl/docs/something/else',
    );
    expect(createToUrl('/baseUrl/', '/docs/something/else/')).toBe(
      '/baseUrl/docs/something/else/',
    );
    expect(createToUrl('/baseUrl/', 'docs/something/else')).toBe(
      'docs/something/else',
    );
    expect(createToUrl('/baseUrl/', './docs/something/else')).toBe(
      './docs/something/else',
    );
    expect(createToUrl('/baseUrl/', 'https://docs/something/else')).toBe(
      'https://docs/something/else',
    );
  });
});

describe('toRedirectFiles', () => {
  it('creates appropriate metadata absolute url', () => {
    const pluginContext = {
      outDir: '/tmp/someFixedOutDir',
      baseUrl: '/',
    };

    const redirectFiles = toRedirectFiles(
      [
        {from: '/abc', to: 'https://docusaurus.io/'},
        {from: '/def', to: 'https://docusaurus.io/docs/intro?a=1'},
        {from: '/ijk', to: 'https://docusaurus.io/docs/intro#anchor'},
      ],
      pluginContext,
      undefined,
    );

    expect(redirectFiles.map((f) => f.fileAbsolutePath)).toEqual([
      path.join(pluginContext.outDir, '/abc/index.html'),
      path.join(pluginContext.outDir, '/def/index.html'),
      path.join(pluginContext.outDir, '/ijk/index.html'),
    ]);

    expect(redirectFiles.map((f) => f.fileContent)).toMatchSnapshot(
      'fileContent',
    );
  });

  it('creates appropriate metadata trailingSlash=undefined', () => {
    const pluginContext = {
      outDir: '/tmp/someFixedOutDir',
      baseUrl: 'https://docusaurus.io',
    };

    const redirectFiles = toRedirectFiles(
      [
        {from: '/abc.html', to: '/abc'},
        {from: '/def', to: '/def.html'},
        {from: '/xyz', to: '/'},
      ],
      pluginContext,
      undefined,
    );

    expect(redirectFiles.map((f) => f.fileAbsolutePath)).toEqual([
      path.join(pluginContext.outDir, '/abc.html/index.html'),
      path.join(pluginContext.outDir, '/def/index.html'),
      path.join(pluginContext.outDir, '/xyz/index.html'),
    ]);

    expect(redirectFiles.map((f) => f.fileContent)).toMatchSnapshot(
      'fileContent',
    );
  });

  it('creates appropriate metadata trailingSlash=true', () => {
    const pluginContext = {
      outDir: '/tmp/someFixedOutDir',
      baseUrl: 'https://docusaurus.io',
    };

    const redirectFiles = toRedirectFiles(
      [
        {from: '/abc.html', to: '/abc'},
        {from: '/def', to: '/def.html'},
        {from: '/xyz', to: '/'},
      ],
      pluginContext,
      true,
    );

    expect(redirectFiles.map((f) => f.fileAbsolutePath)).toEqual([
      path.join(pluginContext.outDir, '/abc.html/index.html'),
      path.join(pluginContext.outDir, '/def/index.html'),
      path.join(pluginContext.outDir, '/xyz/index.html'),
    ]);

    expect(redirectFiles.map((f) => f.fileContent)).toMatchSnapshot(
      'fileContent',
    );
  });

  it('creates appropriate metadata trailingSlash=false', () => {
    const pluginContext = {
      outDir: '/tmp/someFixedOutDir',
      baseUrl: 'https://docusaurus.io',
    };

    const redirectFiles = toRedirectFiles(
      [
        {from: '/abc.html', to: '/abc'},
        {from: '/def', to: '/def.html'},
        {from: '/xyz', to: '/'},
      ],
      pluginContext,
      false,
    );

    expect(redirectFiles.map((f) => f.fileAbsolutePath)).toEqual([
      // Can't be used because /abc.html already exists, and file/folder can't
      // share same name on Unix!
      // path.join(pluginContext.outDir, '/abc.html/index.html'),
      path.join(pluginContext.outDir, '/abc.html.html'), // Weird but on purpose!
      path.join(pluginContext.outDir, '/def/index.html'),
      path.join(pluginContext.outDir, '/xyz/index.html'),
    ]);

    expect(redirectFiles.map((f) => f.fileContent)).toMatchSnapshot(
      'fileContent',
    );
  });

  it('creates appropriate metadata for root baseUrl', () => {
    const pluginContext = {
      outDir: '/tmp/someFixedOutDir',
      baseUrl: '/',
    };
    const redirectFiles = toRedirectFiles(
      [{from: '/abc.html', to: '/abc'}],
      pluginContext,
      undefined,
    );
    expect(redirectFiles.map((f) => f.fileContent)).toMatchSnapshot(
      'fileContent baseUrl=/',
    );
  });

  it('creates appropriate metadata for empty baseUrl', () => {
    const pluginContext = {
      outDir: '/tmp/someFixedOutDir',
      baseUrl: '',
    };
    const redirectFiles = toRedirectFiles(
      [{from: '/abc.html', to: '/abc'}],
      pluginContext,
      undefined,
    );
    expect(redirectFiles.map((f) => f.fileContent)).toMatchSnapshot(
      'fileContent baseUrl=empty',
    );
  });
});

describe('writeRedirectFiles', () => {
  it('write the files', async () => {
    const outDir = `/tmp/docusaurus_tests_${Math.random()}`;

    const filesMetadata = [
      {
        fileAbsolutePath: path.join(outDir, 'someFileName'),
        fileContent: 'content 1',
      },
      {
        fileAbsolutePath: path.join(outDir, '/some/nested/filename'),
        fileContent: 'content 2',
      },
    ];

    await writeRedirectFiles(filesMetadata);

    await expect(
      fs.readFile(filesMetadata[0]!.fileAbsolutePath, 'utf8'),
    ).resolves.toBe('content 1');

    await expect(
      fs.readFile(filesMetadata[1]!.fileAbsolutePath, 'utf8'),
    ).resolves.toBe('content 2');
  });

  it('avoid overwriting existing files', async () => {
    const outDir = `/tmp/docusaurus_tests_${Math.random()}`;

    const filesMetadata = [
      {
        fileAbsolutePath: path.join(outDir, 'someFileName/index.html'),
        fileContent: 'content 1',
      },
    ];

    await fs.outputFile(
      filesMetadata[0]!.fileAbsolutePath,
      'file already exists!',
    );

    await expect(
      writeRedirectFiles(filesMetadata),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"The redirect plugin is not supposed to override existing files."`,
    );
  });
});
