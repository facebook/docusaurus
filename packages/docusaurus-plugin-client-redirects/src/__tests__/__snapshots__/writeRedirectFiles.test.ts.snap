// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toRedirectFiles creates appropriate metadata absolute url: fileContent 1`] = `
[
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/">
    <link rel="canonical" href="https://docusaurus.io/" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/' + window.location.search + window.location.hash;
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/docs/intro?a=1">
    <link rel="canonical" href="https://docusaurus.io/docs/intro?a=1" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/docs/intro?a=1';
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/docs/intro#anchor">
    <link rel="canonical" href="https://docusaurus.io/docs/intro#anchor" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/docs/intro#anchor';
  </script>
</html>",
]
`;

exports[`toRedirectFiles creates appropriate metadata for empty baseUrl: fileContent baseUrl=empty 1`] = `
[
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=/abc">
    <link rel="canonical" href="/abc" />
  </head>
  <script>
    window.location.href = '/abc' + window.location.search + window.location.hash;
  </script>
</html>",
]
`;

exports[`toRedirectFiles creates appropriate metadata for root baseUrl: fileContent baseUrl=/ 1`] = `
[
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=/abc">
    <link rel="canonical" href="/abc" />
  </head>
  <script>
    window.location.href = '/abc' + window.location.search + window.location.hash;
  </script>
</html>",
]
`;

exports[`toRedirectFiles creates appropriate metadata trailingSlash=false: fileContent 1`] = `
[
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/abc">
    <link rel="canonical" href="https://docusaurus.io/abc" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/abc' + window.location.search + window.location.hash;
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/def.html">
    <link rel="canonical" href="https://docusaurus.io/def.html" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/def.html' + window.location.search + window.location.hash;
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/">
    <link rel="canonical" href="https://docusaurus.io/" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/' + window.location.search + window.location.hash;
  </script>
</html>",
]
`;

exports[`toRedirectFiles creates appropriate metadata trailingSlash=true: fileContent 1`] = `
[
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/abc">
    <link rel="canonical" href="https://docusaurus.io/abc" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/abc' + window.location.search + window.location.hash;
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/def.html">
    <link rel="canonical" href="https://docusaurus.io/def.html" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/def.html' + window.location.search + window.location.hash;
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/">
    <link rel="canonical" href="https://docusaurus.io/" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/' + window.location.search + window.location.hash;
  </script>
</html>",
]
`;

exports[`toRedirectFiles creates appropriate metadata trailingSlash=undefined: fileContent 1`] = `
[
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/abc">
    <link rel="canonical" href="https://docusaurus.io/abc" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/abc' + window.location.search + window.location.hash;
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/def.html">
    <link rel="canonical" href="https://docusaurus.io/def.html" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/def.html' + window.location.search + window.location.hash;
  </script>
</html>",
  "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=https://docusaurus.io/">
    <link rel="canonical" href="https://docusaurus.io/" />
  </head>
  <script>
    window.location.href = 'https://docusaurus.io/' + window.location.search + window.location.hash;
  </script>
</html>",
]
`;
